src/main/kotlin/net/thoughtmachine/please/plugin/PleaseLanguage.kt

```diff
@@ -70,54 +83,20 @@ class PleaseFile(viewProvider: FileViewProvider, private var type : PleaseFileTy
     /**
      * Gets the Please package name for the File by walking up the file tree to find the .plzconfig.
      */
-    fun getPleasePackage() : String? {
+    fun getPleasePackage() : Package? {
         // Build definitions file don't belong to a package.
         if (type == PleaseBuildDefFileType) {
             return null
         }
 
-        locatePleaseRepo()
-        return pkg
-    }
-
-    fun getProjectRoot() : Path? {
-        locatePleaseRepo()
-        return repo
-    }
-
-    private fun locatePleaseRepo() {
-        // TODO(jpoole): move build defs out into their own file type
-        if (locatedRepoRoot) {
-            return
-        }
-        var dir = Path.of(virtualFile.path).parent
-        val path = mutableListOf<String>()
-        while(true) {
-            if(dir == null){
-                return
-            }
-
-            val dirFile = dir.toFile()
-            if (dir.toFile().list()?.find { it == ".plzconfig" } != null) {
-                pkg = path.joinToString("/")
-                repo = dir.toAbsolutePath()
-                locatedRepoRoot = true
-                return
-            } else {
-                path.add(0, dirFile.name)
-                dir = dir.parent
-            }
-        }
+        val (root, pkg) = PackageIndexer.forFile(this) ?: return null
+        return PackageService.resolvePackage(project, root, pkg)
```

Comment:
This is the crux of the PR. We now have a `PackagIndexer` which stores the Please root and package path for each BUILD file. We can then look this up via the `PackageService` to get the fully realised package, including if a target is a test, or binary, and what labels it's been tagged with.